### PR TITLE
Report swift-proposal changes

### PR DESCRIPTION
**Pull request by MahdiBM**
Reports new swift proposals or those that have had a change of status.
Uses this url to get the data in JSON: https://download.swift.org/swift-evolution/proposals.json
The whole implementation is rather simple. Get the proposals through the link, check if there has been a change, then report.

* [x] Implemented
* [x] Tests

### CODE REVIEW

File: CODE/Sources/PennyBOT/+Array.swift [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+        while let element = iterator.next() {
+            if try await isInLhs(element) {
+                lhs.append(element)
+            } else {
+                rhs.append(element)
+            }
+        }
+
+        return (Array(lhs), Array(rhs))
+    }
```
Comment (gwynne) [inline CODE/Sources/PennyBOT/+Array.swift:20]:
```suggestion
extension Array {
    func divided(
        _ isInLhs: (Element) async throws -> Bool
    ) async rethrows -> (lhs: [Element], rhs: [Element]) {
        var lhs = [Element]()
        var rhs = [Element]()
        lhs.reserveCapacity(self.count)
        rhs.reserveCapacity(self.count)

        for element in self {
            if try await isInLhs(element) {
                lhs.append(element)
            } else {
                rhs.append(element)
            }
        }
        return (lhs, rhs)
    }
```

File: CODE/Sources/PennyBOT/+Array.swift [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            if try isInLhs(element) {
+                lhs.append(element)
+            } else {
+                rhs.append(element)
+            }
+        }
+
+        return (Array(lhs), Array(rhs))
+    }
+}
```
Comment (gwynne) [inline CODE/Sources/PennyBOT/+Array.swift:40]:
```suggestion
    func divided(
        _ isInLhs: (Element) throws -> Bool
    ) rethrows -> (lhs: ArraySlice<Element>, rhs: ArraySlice<Element>) {
        var copy = self
        let firstOfRhs = try copy.partition(by: isInLhs)
        return (copy[copy.startIndex ..< firstOfRhs], copy[firstOfRhs ..< copy.endIndex])
    }
}
```

File: CODE/Sources/PennyBOT/ProposalsChecker.swift [hunk 3]
```diff
@@ -0,0 +1,191 @@
+import Logging
+import DiscordModels
+import PennyModels
+import Foundation
+
+actor ProposalsChecker {
+    var previousProposals = [Proposal]()
+    var isFirstRound = true
```
Comment (gwynne) [inline CODE/Sources/PennyBOT/ProposalsChecker.swift:8]:
```suggestion
```

File: CODE/Sources/PennyBOT/ProposalsChecker.swift [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+    func check() async throws {
+        let proposals = try await self.proposalsService.get()
+
+        /// If it's the first round, we have nothing to compare to
+        if isFirstRound {
+            self.previousProposals = proposals
+            isFirstRound = false
+            return
+        }
```
Comment (gwynne) [inline CODE/Sources/PennyBOT/ProposalsChecker.swift:45]:
```suggestion
        guard !self.previousProposals.isEmpty else {
            self.previousProposals = proposals
            return
        }
```

**Review by gwynne — APPROVED** [review]
I made a couple suggestions, but they're basically stylistic/idiom nits, nothing functional.

File: CODE/Sources/PennyBOT/+Array.swift (hunk 1 above)
Comment (MahdiBM) [reply CODE/Sources/PennyBOT/+Array.swift:20]: thanks!

File: CODE/Sources/PennyBOT/+Array.swift (hunk 2 above)
Comment (MahdiBM) [reply CODE/Sources/PennyBOT/+Array.swift:39]: thanks!

File: CODE/Sources/PennyBOT/ProposalsChecker.swift (hunk 4 above)
Comment (MahdiBM) [reply CODE/Sources/PennyBOT/ProposalsChecker.swift:45]: Thanks, i'll add this change with the next cleanup commit that's incoming

Comment (MahdiBM) [reply CODE/Sources/PennyBOT/ProposalsChecker.swift:45]: Though i have something against `guard !thing else { }`, so i'll just use `if thing { }` 😅

File: CODE/Sources/PennyBOT/+Array.swift (hunk 2 above)
Comment (MahdiBM) [reply CODE/Sources/PennyBOT/+Array.swift]: I think this one is causing some problems 👀 tests crashing now

**Pull request merged**